stdlib/public/core/String.swift

```diff
@@ -982,11 +982,105 @@ extension String {
 
 extension _StringGutsSlice {
   internal func _withNFCCodeUnits(_ f: (UInt8) throws -> Void) rethrows {
-    for scalar in Substring(self).nfc {
-      try scalar.withUTF8CodeUnits {
-        for byte in $0 {
+    // Fast path: If we're already NFC (or ASCII), then we don't need to do
+    // anything at all.
+    if _guts.isNFC {
```

Comment:
Something to consider for patterns like this is splitting the function in two, so that the fast path can be inlined into callers. I don't know what the state of partial inlining in swiftc is but I haven't found it to be something I can rely on.